src/document/doc-utils.js

```diff
@@ -219,6 +219,37 @@ function stripTrailingHardline(doc, withInnerParts = false) {
   return doc;
 }
 
+function normalizeParts(parts) {
+  const newParts = [];
+
+  const restParts = parts.slice();
+  while (restParts.length !== 0) {
+    const part = restParts.shift();
+
+    if (!part) {
+      continue;
+    }
+
+    if (part.type === "concat") {
+      restParts.unshift(...part.parts);
+      continue;
+    }
+
+    if (
+      newParts.length !== 0 &&
+      typeof newParts[newParts.length - 1] === "string" &&
+      typeof part === "string"
+    ) {
+      newParts.push(newParts.pop() + part);
+      continue;
+    }
+
+    newParts.push(part);
+  }
+
+  return newParts;
+}
```

Comment:
We export this file to plugins, so I think we should documented it

Reply:
https://github.com/prettier/prettier/blob/cedb1a2c2539fcc5813bc4aa4169401d5ea2d7cd/src/index.js#L50
https://github.com/prettier/prettier/blob/cedb1a2c2539fcc5813bc4aa4169401d5ea2d7cd/src/document/index.js

Reply:
I was wrong, we don't have documentation about utils

Reply:
This is definitely out of scope for this PR. The entire `prettier.doc` namespace isn't documented.